File: py/pre.py
# Copyright (c) 2013, Charles Childers
#
# pre is the "parable runtime", a quick tool for running a parable
# program and displaying the results.
# ===================================================================

# Dependencies
import sys
from parable import *

def display_stack():
    global stack, types
    i = 0
    while i < len(stack):
        if i == len(stack) - 1:
            sys.stdout.write("TOS\t" + str(i))
        else:
            sys.stdout.write("\t" + str(i))

        if types[i] == TYPE_NUMBER:
            sys.stdout.write("\t#" + str(stack[i]))
        elif types[i] == TYPE_CHARACTER:
            sys.stdout.write("\t$" + str(chr(stack[i])))
        elif types[i] == TYPE_STRING:
            sys.stdout.write("\t'" + slice_to_string(stack[i]) + "'")
            sys.stdout.write("\n\t\tstored at: " + str(stack[i]))
        elif types[i] == TYPE_FUNCTION:
            sys.stdout.write("\t&" + str(stack[i]))
            if pointer_to_name(stack[i]) != "":
                sys.stdout.write("\n\t\tpointer to: " + pointer_to_name(stack[i]))
        elif types[i] == TYPE_FLAG:
            if stack[i] == -1:
                sys.stdout.write("\ttrue")
            elif stack[i] == 0:
                sys.stdout.write("\tfalse")
            else:
                sys.stdout.write("\tmalformed flag")
        else:
            sys.stdout.write("\tunmatched type on stack!")
        sys.stdout.write("\n")
        i += 1


def display_errors():
    for e in errors:
        sys.stdout.write("\n" + e)


def display():
    display_stack()
    display_errors()
    clear_errors()


def load_file(file):
    f = open(file).readlines()
    for line in f:
        if len(line) > 1:
            s = compile(line, request_slice())
            interpret(s)


if __name__ == '__main__':
    prepare_slices()
    prepare_dictionary()
    parse_bootstrap(open('bootstrap.p').readlines())
    collect_unused_slices()

    load_file(sys.argv[1])
    display()
    sys.stdout.flush()

File: php/pre.php
<?
require_once("parable.php");

$bootstrap = file($argv[1]);

foreach ($bootstrap as $src)
{
    if (strlen(trim($src)) > 0)
    {
        $s = compile($src, request_slice());
        interpret($s);
    }

    foreach ($errors as $err)
    {
        echo "$err";
    }
    $errors = array();
}



$TYPE_NUMBER    = 100;
$TYPE_STRING    = 200;
$TYPE_CHARACTER = 300;
$TYPE_FUNCTION  = 400;
$TYPE_FLAG      = 500;

$i = 0;
while ($i < count($stack))
{
    if ($types[$i] == $TYPE_NUMBER)
    {
        echo $i, "\t", ' #', $stack[$i];
    }
    if ($types[$i] == $TYPE_CHARACTER)
    {
        echo $i, "\t", ' $', chr($stack[$i]);
    }
    if ($types[$i] == $TYPE_STRING)
    {
        echo $i, "\t", " '", slice_to_string($stack[$i]), "'";
    }
    if ($types[$i] == $TYPE_FUNCTION)
    {
        echo $i, "\t", ' &', $stack[$i];
    }
    if ($types[$i] == $TYPE_FLAG)
    {
        echo $i, "\t", ' FLAG: ', $stack[$i];
    }
    echo "\n";
    $i += 1;
}

?>
